### ISSUE TITLE
Name for the process argument: imagery / collections / process / ...

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-api-poc#49]
Most processes usually have an input argument that contains results from the previous processes or image collections (see ttps://open-eo.github.io/openeo-api-poc/processgraphs/index.html). Currently `imagery` is used quite often, but it was already called differently before (`collections`).

Important to note is that this term depends on the process definition itself (e.g. `filter_bands`). It is not part of the process graph definition itself. The underlying idea from @appelmar, @flahn and me was: Processes that expect images should use `imagery` (or similar), but other processes can use different names. For example, if there is a process that needs a timeseries as input, the argument can also be named accordingly as `timeseries` (or similar).

@aljacob understandably voted to discuss the names we use to define the processes.

I personally prefer to use `imagery` instead of `collections` for image-based input due to the fact that `imagery` is more precise. `collections` can be anything and users might think that an array like data type is expected as input, which is not the case. `imagecollection` would be precise, but is basically a synonym for `imagery`.
I have not yet thought of other data types and their names, e.g. for `timeseries` or whatever we might want to process.

**Comment by GreatEmerald** [comment]
What if a process can take either as an input? I can see `filter_daterange` being valid for both image collections and timeseries.

One other option would be something more generic, like `input`. It's not a terribly exciting name, but in every case the first argument is the input for the process, whatever form that may be, and the other arguments tell the process what to do with that input.

This also relates to #47, since the idea to call the argument differently depending on the type it expects would not be necessary if we had type definitions. Parameter `input` with type `timeseries` is pretty self-evident and would simplify making clients, since they can continue to assume the first argument is going to be the input.

**Comment by aljacob** [comment]
@m-mohr & @GreatEmerald I was under the impression that a **collection** could be a **timeseries** anyways, so they have not been different things  in my mind. I guess it boils down to, what anyone of us is used to based on the development environment and software packages that we typically use. E.g. I hardly ever work on individual images, but rather on a time series, so when thinking of a **collection** of images I automatically think of what I guess you denote as **timeseries** and what you denote as **collection** I probably see as an **imagestack**, meaning a stack of different bands or other superimposed raster data on the same location. The term collection for me is agnostic to if it is stacked in time or feature space and hence a more general choice, just like **input** would be, although arguable any scalar type of inputs such as the wavelength to filter for or similar could also be seen as **input**.

**Comment by edzer** [comment]
We need the glossary!

**Comment by m-mohr** [comment]
This is actually part of the openEO data set and process descriptions milestone (v0.4). The naming is not part of the API at all, but part of the process definitions and will be not relevant at all if we strictly follow the idea behind implementing #47.

**Comment by GreatEmerald** [comment]
Indeed, however, that is for the most part pushing the issue to a different level of implementation. If the definition of all processes starts with an argument called "imagery", we're effectively back to square one.

Of course, this does allow naming the first argument differently for different processes. But that still needs to be decided, now for each process. Also, one reason why "imagery" is not a good name is because it's actually "imagery or process". Granted, you can see "process" as "modified imagery", because there's probably imagery at the end of the chain, though you can never be sure unless you type check the whole chain. Without that, you could pass a timeseries-type object to an argument called "imagery", and it would still be valid as far as the API is concerned.

Also, this argument is used in each client as a pipeline element, and pipelines assume that you pipe the first argument of a function, as opposed to checking types etc. So in that sense the name doesn't matter at all because the users don't see it most of the time. But it's important that it is the first argument, always. So from that perspective, might as well call it "process" instead of "imagery". The only thing the name changes is how the output process graph JSON looks like. And in almost all cases (or all cases altogether as per #56), the argument is a process. It is quite confusing to see the process graph JSON with a whole lot of "imagery" levels, when in reality these are just chained processes, with the imagery (or data) itself being at the very end of the chain.

**Comment by m-mohr** [comment]
After thinking about this for quite some time now and all your good input, I think it is time to wrap this up and come to a conclusion.

So, what questions have arised?
The main question is: How do we inject the data a process is working on?
More specifically:

1. Is it always the first argument?
2. Are there multiple arguments if multiple data sets are used, e.g. for a union operation?
3. What is actually the data type of the argument?
4. How is the argument named?

---

1. The JSON spec doesn’t require to keep the order of object keys. Therefore we can't require it to be the first. We can just work on the key names. So ultimately clients would need to parse the process descriptions to know where to put the data. This was my first idea, but seems to be overly complicated and over-engineered. Therefore we can only use a standardized name to make it simple and will not rely on the ordering, i.e. it is not necessarily the first argument in the process graph, BUT can/should be in the client libraries.
2. As we go for simplicity and therefore for the named argument, we should define it to take either an array of input data sets. We can make a shortcut and also only allow a single data set. Therefore union would simply take an array. Arrays - in contrast to objects - are ordered and therefore processes could use this to specify a specific meaning to the data based on the order of the data sets.
3. Probably something proprietary based on what the back-end is working with. We may just call it "EOData" or so...
4. I'd suggest to just name it the same way we name our "data type" in (3). 

Coming back to naming. This issue already has the following suggestions and I also add some more:
* collection(s)
* ~imagery~: As imagery isn't covering all use cases (e.g. timeseries) this can be safely excluded, I think.
* process: Process seems a little to ambiguous. We don't really inject a process here, but mostly but not always processed data.
* input: Seems a little to generic, other arguments are also input for the process.
* eo_data: Actually my favorite for now.
* data_cube

Opinions on the naming? 

Other than that, I'd suggest to allow using either a single element or an array as input. That depends a little on #56. If we want the process graph to be "flat" than we should allow a single element. Otherwise we could also always wrap it in an array.

**Comment by m-mohr** [comment]
We discussed that again in the dev telco and we couldn't find a solution for this. 
It was discussed that having it in the first parameter only may not be good enough for some use cases. and we still may need multiple arguments, i.e. for Sampling a raster time series with a feature collection. In this case it should be made clear what is what and not merged into one list. Still, without lists it may be a problem when we want to specify an arbitrary number of datasets, e.g. a union of five datasets.
There is still much to discuss (in person?) and no solution found yet.

**Comment by GreatEmerald** [comment]
The point of for having it as the first argument is for making it easy to chain in clients (pipes can be used). At the same time, the clients' API does not necessarily have to be the same as the core API, so clients could still use piping, but resolve what the pipe means on an individual process basis. And when we have two inputs, we can't use piping effectively anyway.

I think the union example is nice and should be fleshed out a bit more. An example in JSON would help. It's true that it's nice when the union can accept an arbitrary number of data to union (with an array), but on the other hand that could also be achieved by chaining a whole bunch of pair unions together. And which one is preferred might depend on the backend...

While `eo_data` is a nice name, it still bothers me that it's not really data as much as it is a process graph fragment. It's data that potentially has been modified (on the fly). It's still data in the end, of course, but it just seems misleading when looking at JSON process graphs, because that block contains much more than data... That's why my suggestion was for `input`.

**Comment by m-mohr** [comment]
Postponed to v0.4 as in that version the processes will be "finalized" and it doesn't make sense to introduce an intermediate step in v0.3, which may be changed again in v0.4.
